compiler/src/iree/compiler/Codegen/Dialect/PCF/IR/PCFOps.cpp

```diff
@@ -0,0 +1,1085 @@
+}
+
+ValueRange GenericOp::getResultDims(int64_t i) {
+  if (getIsTied()[i]) {
+    return {};
+  }
+
+  int64_t startIndex = 0;
+  for (auto [curr, isTied] : llvm::enumerate(getIsTied())) {
+    if (curr == i) {
+      break;
+    }
+    if (!isTied) {
+      startIndex += getResultType(curr).getNumDynamicDims();
+    }
+  }
+
+  return ValueRange(getDynamicSizes().slice(
+      startIndex, startIndex + getResultType(i).getNumDynamicDims()));
+}
+
+//===----------------------------------------------------------------------===//
+// LoopOp
+//===----------------------------------------------------------------------===//
+
+void LoopOp::getAsmBlockArgumentNames(Region &region,
+                                      OpAsmSetValueNameFn setNameFn) {
+  for (Value v : getIdArgs()) {
+    setNameFn(v, "id");
+  }
+  for (Value v : getRegionRefArgs()) {
+    setNameFn(v, "ref");
+  }
+}
+
+LogicalResult LoopOp::verify() {
+  if (getCount().empty()) {
+    return emitOpError("expected at least one iteration count argument");
+  }
+  if (getBody()->getNumArguments() < getNumIdArgs()) {
+    return emitOpError("fewer body arguments than specified number of ids.");
+  }
+  return verifyParallelBodyOp(*this, /*numLeadingArgs=*/0, getNumIdArgs(),
+                              getIdArgs());
+}
+
+void LoopOp::build(mlir::OpBuilder &b, mlir::OperationState &result,
+                   ScopeAttr scope, ValueRange count, bool syncOnReturn) {
+  LoopOp::build(b, result, TypeRange(), scope, count, ArrayRef<Value>{},
+                ArrayRef<Value>{}, ArrayRef<bool>{}, syncOnReturn);
+}
+
+void LoopOp::build(mlir::OpBuilder &b, mlir::OperationState &result,
+                   ScopeAttr scope, ValueRange count, ValueRange inits,
+                   bool syncOnReturn) {
+  SmallVector<bool> isTied(inits.size(), true);
+  SmallVector<Type> resultTypes =
+      llvm::map_to_vector(inits, [](Value v) -> Type { return v.getType(); });
+  LoopOp::build(b, result, resultTypes, scope, count, inits, ArrayRef<Value>{},
+                isTied, syncOnReturn);
+}
+
+void LoopOp::build(mlir::OpBuilder &b, mlir::OperationState &result,
+                   TypeRange resultTypes, ScopeAttr scope, ValueRange count,
+                   ValueRange dynamicSizes, bool syncOnReturn) {
+  SmallVector<bool> isTied(resultTypes.size(), false);
+  LoopOp::build(b, result, resultTypes, scope, count, ArrayRef<Value>{},
+                dynamicSizes, isTied, syncOnReturn);
+}
+
+void LoopOp::build(mlir::OpBuilder &b, mlir::OperationState &result,
+                   TypeRange resultTypes, ScopeAttr scope, ValueRange count,
+                   ValueRange inits, ValueRange dynamicSizes,
+                   ArrayRef<bool> isTied, bool syncOnReturn) {
+
+  result.addAttribute(LoopOp::getScopeAttrName(result.name), scope);
+  result.addOperands(count);
+  result.addOperands(inits);
+  result.addOperands(dynamicSizes);
+  result.addTypes(resultTypes);
+
+  result.addAttribute(
```

Comment:
`result.getOrAddProperties().operandSegmentSizes - you don't need to make this an attribute, that just costs